snapcraft/models/project.py

```diff
@@ -1261,6 +1290,9 @@ def get_build_plan(self) -> List[BuildInfo]:
         # set default value
         if self.platforms is None:
             self.platforms = {get_host_architecture(): None}
+            # For backwards compatibility with core22, convert the platforms.
+            if self.base == "core22" and self.architectures:
```

Comment:
> self.architectures

I'm thinking of the other cases that we support for `core22`. Correct me if I'm wrong, but for core22 snaps:

- the `architectures` keyword isn't even [needed](https://github.com/canonical/snapcraft/blob/7.5.5/snapcraft/projects.py#L451) right?
- some fields [accept](https://github.com/canonical/snapcraft/blob/7.5.5/snapcraft/projects.py#L92) `"all"`, which at some point gets converted into the host architecture (I think?), but our notion of BuildInfo can't handle that (can it)?

I think these are some good test cases to add, at least

Reply:
That's a good point. What *should* be the default behaviour if there are no `architectures` and no `platforms` provided? Right now it's just "build on the host," but that's probably not what we want for remote builds. Right now this works without an `architectures` keyword, but it probably does the wrong thing in the `remote-build` case since it ends up building on just the client's architecture. Maybe we make the build plan just list all supported architectures and not do cross-compiling? Filtering the plan will result in the same behaviour for local builds.

I'll have to give a bit more thought to the second point and how we want this to behave. @mr-cal wdyt?

Reply:
I'm not sure what to do about `build-for: all` for remote-builds.  There is no precedent because it's never worked ([1](https://bugs.launchpad.net/snapcraft/+bug/1992557) [2](https://bugs.launchpad.net/snapcraft/+bug/2007789) [3](https://github.com/canonical/snapcraft/issues/4144)).  This may be a good time for snapcraft to raise an error when converting `build-for: [all]` to `platforms`.

Regarding the behavior when no `platforms` or `architectures` are defined.  Building for the client architecture makes sense if remote-build is used for development and testing.  Building for each architecture makes sense if remote-build is used for CI and distribution.  I lean towards the former because `snapcraft pack` and `snapcraft remote-build` would both produce 1 snap, built on and built for the client's architecture.

Reply:
OK, so I've left the default build plan as creating for only the current arch, and I've set it to error out explicitly if there's a `build-for: all` for core22. I've also made `from_architectures` explicitly raise an exception if an architecture has `build-for: all`